### PR TITLE
Fixes in GRPO recipe and README

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

- change the recipe with lower max_target_length
- Fix the README with more information on inference_rollouts
- update pathwaysutils installation in all requirements.txt files

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

https://cloudlogging.app.goo.gl/RtuoGL4Bd7LRBBWEA

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/experimental/rl/grpo_inference.yml [hunk 1]
```diff
@@ -22,3 +22,19 @@ return_log_prob: True
...
+sa_block_kv_compute: 128
+sa_block_q_dkv: 128
+sa_block_kv_dkv: 128
+sa_block_kv_dkv_compute: 128
+sa_block_q_dq: 128
+sa_block_kv_dq: 128
+sa_use_fused_bwd_kernel: False
+sa_q_layout: "HEAD_DIM_MINOR"
+sa_k_layout: "HEAD_DIM_MINOR"
+sa_v_layout: "HEAD_DIM_MINOR"
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo_inference.yml]: nit: please add the new line

File: MaxText/experimental/rl/grpo.yml [hunk 2]
```diff
@@ -32,3 +32,19 @@ return_log_prob: True
...
+sa_block_kv_compute: 128
+sa_block_q_dkv: 128
+sa_block_kv_dkv: 128
+sa_block_kv_dkv_compute: 128
+sa_block_q_dq: 128
+sa_block_kv_dq: 128
+sa_use_fused_bwd_kernel: False
+sa_q_layout: "HEAD_DIM_MINOR"
+sa_k_layout: "HEAD_DIM_MINOR"
+sa_v_layout: "HEAD_DIM_MINOR"
```
Comment (A9isha) [inline MaxText/experimental/rl/grpo.yml]: nit: please add the new line

**Review by A9isha — APPROVED** [review]

File: MaxText/experimental/rl/README.md [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
+*   **FSDP (Fully Sharded Data Parallelism):** A technique for distributing model parameters across multiple devices during training, allowing for efficient training of large models.
 *   **DP (Data Parallelism):** A technique for distributing data across multiple devices during inference, allowing for efficient inference of large models.
-*   **Pathways:** A system that allows to orchestrate computations across multiple TPU slices, enabling training and inference on very large TPU topologies.
+*   **Sharding:** For more information on sharding, refer to [getting_started/Sharding.md](https://github.com/AI-Hypercomputer/maxtext/blob/main/getting_started/Sharding.md)
+*   **[Pathways](https://cloud.google.com/ai-hypercomputer/docs/workloads/pathways-on-cloud/pathways-intro):** A system that allows to orchestrate computations across multiple TPU slices, enabling training and inference on very large TPU topologies.
+*   **inference_rollouts:** The frequency in training steps to reshard and copy training parameters to the inference mesh. Must be an integer between 1 and `steps`.
```
Comment (shralex) [inline MaxText/experimental/rl/README.md]: This is different from previous concepts above in that its a config parameter. I suggest to start the sentence with that info. Also, what is steps ? is that STEPS ? is that the total number of training steps to execute ?

File: MaxText/experimental/rl/grpo.yml [hunk 4]
```diff
@@ -32,3 +32,19 @@ return_log_prob: True
 
 add_bos: False
 add_eos: False
+
+### Splash attention block sizes
+# These can be tuned for specific hardware generations, and can be set up to
+# the model's sequence length.
```
Comment (shralex) [inline MaxText/experimental/rl/grpo.yml]: are the values below tuned for some hardware or just examples ?

File: MaxText/experimental/rl/grpo_inference.yml [hunk 5]
```diff
@@ -22,3 +22,19 @@ return_log_prob: True
 
 add_bos: False
 add_eos: False
+
+### Splash attention block sizes
+# These can be tuned for specific hardware generations, and can be set up to
```
Comment (shralex) [inline MaxText/experimental/rl/grpo_inference.yml]: same question here, plus would be good to point out any differences from grpo.yml settings or guidelines if we have them

File: requirements_with_jax_ai_image.txt [hunk 6]
```diff
@@ -19,7 +19,7 @@ transformers
 mlperf-logging@git+https://github.com/mlperf/logging.git
 google-jetstream@git+https://github.com/AI-Hypercomputer/JetStream.git
 jsonlines
-pathwaysutils==0.1.1
+pathwaysutils@git+https://github.com/AI-Hypercomputer/pathways-utils.git
```
Comment (bvandermoon) [inline requirements_with_jax_ai_image.txt:22]: I think this is not intended for GPUs. But just want to confirm this does not need to be added to `constraints_gpu` also

File: pytest.ini [hunk 7]
```diff
@@ -10,6 +10,8 @@ addopts =
     --ignore=MaxText/tests/train_int8_smoke_test.py
     --ignore=MaxText/tests/train_gpu_smoke_test.py
     --ignore=MaxText/tests/train_using_ragged_dot_smoke_test.py
+    --ignore=MaxText/tests/grpo_trainer_correctness_test.py
+    --ignore=MaxText/tests/offline_engine_test.py
```
Comment (bvandermoon) [inline pytest.ini:14]: Why are we ignoring these tests?

File: MaxText/experimental/rl/grpo_trainer.py [hunk 8]
```diff
@@ -639,6 +639,7 @@ def generation_worker_fn(
   generation_thread.start()
 
   try:
+    last_step_completion = datetime.datetime.now()
```
Comment (bvandermoon) [inline MaxText/experimental/rl/grpo_trainer.py:642]: Does this + the timing below work properly without `jax.block_until_ready(...)`?

**Review by bvandermoon — COMMENTED** [review]
Generally LGTM once @shralex's comments have also been addressed. Mainly just want to make sure the timing around `last_step_completion` is working properly

File: MaxText/experimental/rl/README.md (hunk 3 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/README.md]: I changed the statement

File: MaxText/experimental/rl/grpo.yml (hunk 4 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo.yml]: These are tuned for the example ran from test_grpo.sh. I have now mentioned this in the comments

File: MaxText/experimental/rl/grpo_inference.yml (hunk 5 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_inference.yml]: Added a comment explaining why grpo_inference.yml is needed

File: requirements_with_jax_ai_image.txt (hunk 6 above)
Comment (khatwanimohit) [reply requirements_with_jax_ai_image.txt:22]: I noticed pathways-utils is not added as part of the contraints file. This is fine as that'll always pull from head

File: pytest.ini (hunk 7 above)
Comment (khatwanimohit) [reply pytest.ini:14]: resharding feature from pathwaysutils uses a very recent JAX API. Will enable these tests once new JAX is released.

File: MaxText/experimental/rl/grpo_trainer.py (hunk 8 above)
Comment (khatwanimohit) [reply MaxText/experimental/rl/grpo_trainer.py:642]: Discussed offline. This works as expected because we buffer and write the metrics which will indirectly cause a blocking operation
